src/main/java/org/apache/sysml/runtime/controlprogram/paramserv/LocalAggregationService.java

```diff
@@ -0,0 +1,151 @@
+import org.apache.sysml.runtime.DMLRuntimeException;
+import org.apache.sysml.runtime.controlprogram.FunctionProgramBlock;
+import org.apache.sysml.runtime.controlprogram.context.ExecutionContext;
+import org.apache.sysml.runtime.instructions.cp.CPOperand;
+import org.apache.sysml.runtime.instructions.cp.Data;
+import org.apache.sysml.runtime.instructions.cp.FunctionCallCPInstruction;
+import org.apache.sysml.runtime.instructions.cp.ListObject;
+
+public class LocalAggregationService extends AggregationService implements Runnable {
+
+	protected static final Log LOG = LogFactory.getLog(LocalAggregationService.class.getName());
+	private static final long POLL_FREQUENCY = 1;
+
+	private String _funcNS;
+	private String _funcName;
+	private ArrayList<DataIdentifier> _inputs;
+	private ArrayList<DataIdentifier> _outputs;
+	private CPOperand[] _boundInputs;
+	private ArrayList<String> _inputNames;
+	private ArrayList<String> _outputNames;
+
+	public LocalAggregationService(String aggFunc, ExecutionContext ec, ListObject models, ListObject hyperParams,
+			ParamServer ps, int workerNum) {
+		super(aggFunc, ec, models, hyperParams, ps, workerNum);
+
+		// Fetch the agg function
+		String[] keys = DMLProgram.splitFunctionKey(_aggFunc);
+		_funcName = keys[0];
+		if (keys.length == 2) {
+			_funcNS = keys[0];
+			_funcName = keys[1];
+		}
+		FunctionProgramBlock func = _ec.getProgram().getFunctionProgramBlock(_funcNS, _funcName);
+		_inputs = func.getInputParams();
+		_outputs = func.getOutputParams();
+		_boundInputs = _inputs.stream()
+				.map(input -> new CPOperand(input.getName(), input.getValueType(), input.getDataType()))
+				.toArray(CPOperand[]::new);
+		_inputNames = _inputs.stream().map(DataIdentifier::getName).collect(Collectors.toCollection(ArrayList::new));
+		_outputNames = _outputs.stream().map(DataIdentifier::getName).collect(Collectors.toCollection(ArrayList::new));
+	}
+
+	@Override
+	public void run() {
+
+		// Firstly, start by broadcasting the model
+		pushGlobalParams();
+		ParamservUtils.populate(_ec, _hyperParams, _inputs);
+		ParamservUtils.populate(_ec, _model, _inputs);
+
+		while (isAlive()) {
+			int finished = 0;
+			boolean[] tables = new boolean[_workerNum];
+
+			for (int i = 0; i < _workerNum && finished < _workerNum; i++) {
+				if (tables[i]) {
+					continue;
+				}
+				// Pull the gradients of each worker and update the model
+				Data gradients = _ps.pull(ParamServer.GRADIENTS_PREFIX + i);
+				if (gradients != null) {
+					LOG.info(String.format("Successfully pulled the gradients [size:%d kb] of worker_%d.",
+							((ListObject) gradients).getDataSize() / 1024, i));
+					aggregate((ListObject) gradients);
+					finished++;
+					tables[i] = true;
+				}
+				try {
+					TimeUnit.SECONDS.sleep(POLL_FREQUENCY);
+				} catch (InterruptedException e) {
+					throw new DMLRuntimeException("Aggregation service: Failed to sleep when polling the gradients.",
+							e);
+				}
```

Comment:
It should generally not be necessary to poll anything here. This might be an indicator of a suboptimal design - i.e., maybe by merging the aggregation service with the parameter server (or using it inside the parameter server), we get a leaner design where we could simply trigger aggregation whenever a worker pushed new gradients. Also, for multi-threaded environments this polling can often be replaced via wait/notify.